pkg/actuators/machine/actuator.go

```diff
@@ -300,34 +295,44 @@ func (a *Actuator) CreateMachine(cluster *clusterv1.Cluster, machine *machinev1.
 		}
 	}
 
-	userData := []byte{}
-	if machineProviderConfig.UserDataSecret != nil {
-		var userDataSecret corev1.Secret
-		err := a.client.Get(context.Background(), client.ObjectKey{Namespace: machine.Namespace, Name: machineProviderConfig.UserDataSecret.Name}, &userDataSecret)
-		if err != nil {
-			return nil, a.handleMachineError(machine, apierrors.CreateMachine("error getting user data secret %s: %v", machineProviderConfig.UserDataSecret.Name, err), createEventAction)
-		}
-		if data, exists := userDataSecret.Data[userDataSecretKey]; exists {
-			userData = data
-		} else {
-			glog.Warningf("%s: Secret %v/%v does not have %q field set. Thus, no user data applied when creating an instance.", machine.Name, machine.Namespace, machineProviderConfig.UserDataSecret.Name, userDataSecretKey)
-		}
+	userData, err := a.getUserData(machine, machineProviderConfig)
+	if err != nil {
+		return nil, err
 	}
 
 	instance, err := launchInstance(machine, machineProviderConfig, userData, awsClient)
 	if err != nil {
-		return nil, a.handleMachineError(machine, apierrors.CreateMachine("error launching instance: %v", err), createEventAction)
+		return nil, a.handleMachineError(machine, err, createEventAction)
 	}
 
 	err = a.updateLoadBalancers(awsClient, machineProviderConfig, instance, machine.Name)
 	if err != nil {
-		return nil, a.handleMachineError(machine, apierrors.CreateMachine("error updating load balancers: %v", err), createEventAction)
+		return nil, a.handleMachineError(machine, err, createEventAction)
 	}
 
 	a.eventRecorder.Eventf(machine, corev1.EventTypeNormal, "Created", "Created Machine %v", machine.Name)
 	return instance, nil
 }
 
+func (a *Actuator) getUserData(machine *machinev1.Machine, machineProviderConfig *providerconfigv1.AWSMachineProviderConfig) ([]byte, error) {
+	if machineProviderConfig.UserDataSecret != nil {
```

Comment:
I realize you're just moving this around (which is already an improvement) but can we make this just bail early if the secret is nil? As is, 100% of the function is wrapped in this if statement.